src/manifolds/FixedRankMatrices.jl

```diff
@@ -287,16 +284,18 @@ singular values and $U$ and $V$ are shortened accordingly.
 retract(::FixedRankMatrices, ::Any, ::Any, ::PolarRetraction)
 
 function retract!(
-    ::FixedRankMatrices{M,N,k},
+    ::FixedRankMatrices{m, n, k},
     q::SVDMPoint,
     p::SVDMPoint,
     X::UMVTVector,
     ::PolarRetraction,
-) where {M,N,k}
-    s = svd(p.U * Diagonal(p.S) * p.Vt + (p.U * X.M * p.Vt + X.U * p.Vt + X.U * X.Vt))
-    q.U .= s.U[:, 1:k]
-    q.S .= s.S[1:k]
-    q.Vt .= s.Vt[1:k, :]
+) where {m, n, k}
+    QU, RU = qr([p.U X.U])
+    QV, RV = qr([p.Vt' X.Vt'])
+    T = svd(RU * [diagm(p.S) + X.M I; I zeros(k, k)] * RV')
+    q.U .= QU * T.U[:, 1:k]
+    q.S .= T.S[1:k]
+    q.Vt .= T.Vt[1:k, :] * QV'
```

Comment:
Can you use views here to avoid allocating when you slice?
Also you could use `mul!` to multiply in-place instead of multiplying and then copying. This should reduce the 5 allocations here to 0.
```suggestion
    mul!(q.U, QU, @view(T.U[:, 1:k]))
    q.S .= @view(T.S[1:k])
    mul!(q.Vt, @view(T.Vt[1:k, :]), QV')
```

Reply:
I made the changes. It's indeed allocating less memory and is slightly faster.
However the line
`mul!(q.Vt, @view(T.Vt[1:k, :]), QV')`
throws an error because of supposedly dimension mismatchs.

But the dimensions are right, and actually computing the product `@view(T.Vt[1:k, :]) * QV'` raises no error.
I think it's because `q.Vt` is an Adjoint. If I do
```julia
C = zeros(size(q.Vt))
mul!(C, @view(T.Vt[1:k, :]), QV')
```
then it works.

That's weird.

I guess we could just keep the operation `q.Vt .= T.Vt[1:k, :] * QV'` unless there is an easy fix.

Reply:
When do you get the error? The test cases for FixedRank run fine for me (at least for the retraction).

Reply:
If `q.Vt` is an `Adjoint`, then you can just perform the adjoint operation:
```julia
mul!(q.Vt', QV, @view(T.Vt'[:, 1:k]))
```

@kellertuer is there a reason `Vt` is stored as an `Adjoint`? In an `SVD` object, usually `Vt` is a `Matrix`, and `V` is the `Adjoint`.

Reply:
I am not aware of any reason, I wasn't even aware it's stored as an adjoint, but sure, I usually stored `V` I think. if storing `Vt` is easier, we can surely do that.

Reply:
Unfortunately the adjoint operation
```julia
mul!(q.Vt', QV, @view(T.Vt'[:, 1:k]))
```
doesn't work either. It really looks like a bug to me.

It's an adjoint just because of the way `random_tangent` samples the vector. But we can make it a `Matrix` if you think it's better.

Reply:
I didn't push anything yet kellertuer. That's why you don't see the error. I wanted to make sure that it's ok to keep the last operation as `q.Vt .= T.Vt[1:k, :] * QV'`.

Reply:
If we could resolve this by not storing the adjoint but the actual matrix (as @sethaxen mentioned) then I would maybe opt for that, since that's easier to handle. I don't know why I originally saved the adjoint and not the matrix in the struct, I feel the matrix would be more natural. So if qe could carefully switch to storing a matrix, then the original comment of this change would work again :)

Reply:
I tried without the adjoint. The operation doesn't throw an error anymore. But: it doesn't compute the right result.
An example is better than long sentences.
Try to run this code:
```julia
A = randn(5, 4)
Q, R = qr(A)
B = randn(4, 4)
C = randn(2, 5);

mul!(C, @view(B[1:2, :]), Q')
display(C)

display(@view(B[1:2, :]) * Q')
```
The result is pretty surprising.

It seems like the operation `mul!(C, @view(B[1:2, :]), Q')` doesn't override `C` but takes into account what was written inside.

Reply:
Oh yeah, `qr(A).Q` has really unique behavior because it's an `AbstractMatrix` with a certain size but can be multiplied by matrices of different first dimensions. Probably `Adjoint` here is breaking its behavior for `mul!` only. This works:
`mul!(C, @view(B[1:2, :]), Matrix(Q)')`.

Reply:
Yes this is working!
So the last operation could become
```julia
mul!(q.Vt, @view(T.Vt[1:k, :]), Matrix(QV)')
```
but it's not faster than
```julia
q.Vt .= T.Vt[1:k, :] * QV'
```
(it's actually slower.)

I'll put `mul!(q.Vt, @view(T.Vt[1:k, :]), Matrix(QV)')` for now but this should probably be updated in the future.

Reply:
On my machine, they're equivalent in speed, but the `mul!` one allocates less:
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = randn(5, 4);

julia> Q, R = qr(A);

julia> B = randn(4, 4);

julia> C = randn(2, 5);

julia> foo!(C, B, Q) = mul!(C, @view(B[1:2, :]), Matrix(Q)');

julia> bar!(C, B, Q) = (C .= @view(B[1:2, :]) * Q'; C);

julia> @btime $(foo!)($C, $B, $Q);
  1.442 μs (3 allocations: 688 bytes)

julia> @btime $(bar!)($C, $B, $Q);
  1.444 μs (9 allocations: 784 bytes)
```

It probably doesn't matter, but I think this is fine.

Reply:
Hmm interesting. For small dimensions I have the same result. But when dimensions are larger `bar!` becomes 2 times faster even though it allocates more:
```julia
using LinearAlgebra, BenchmarkTools

n = 1000
m = 800
p = m ÷ 2

A = randn(n, m);
Q, R = qr(A);
B = randn(m, m);
C = randn(p, n);

foo!(C, B, Q) = mul!(C, @view(B[1:p, :]), Matrix(Q)');
bar!(C, B, Q) = (C .= @view(B[1:p, :]) * Q'; C);

@btime $(foo!)($C, $B, $Q);
@btime $(bar!)($C, $B, $Q);

  18.510 ms (14 allocations: 17.09 MiB)
  8.450 ms (20 allocations: 6.10 MiB)
```

Reply:
Oh, good spot, then I'd say `bar!` should be preferred. Also perhaps should use `copyto!` instead of broadcasting when filling the whole array, since it's simpler and more general. so e.g.
```julia
copyto!(q.Vt, @view(T.Vt[1:k, :]) * QV')
```